src/py42/services/_connection.py

```diff
@@ -0,0 +1,258 @@
+from __future__ import print_function
+
+import json as json_lib
+from threading import Lock
+
+from requests.adapters import HTTPAdapter
+from requests.exceptions import HTTPError
+from requests.models import Request
+from requests.sessions import Session
+
+import py42.settings as settings
+from py42._compat import urljoin
+from py42._compat import urlparse
+from py42.exceptions import Py42Error
+from py42.exceptions import Py42FeatureUnavailableError
+from py42.exceptions import raise_py42_error
+from py42.response import Py42Response
+from py42.services._auth import C42RenewableAuth
+from py42.settings import debug
+from py42.util import format_dict
+
+SESSION_ADAPTER = HTTPAdapter(pool_connections=200, pool_maxsize=4, pool_block=True)
+
+ROOT_SESSION = Session()
```

Comment:
This is the one session to rule them all, all requests got through this and simply have different domains / auth passed through it. This being a singleton combined with the session adapter above is essentially what solves the sockets problem.